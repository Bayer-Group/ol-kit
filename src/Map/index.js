/*
  _Map is wrapped by a <Provider> to setup context
  without requiring a declartive <Provider> on the page
*/

export { default as Map } from './_Map'
export * from './utils'
