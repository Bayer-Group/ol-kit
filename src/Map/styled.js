import styled from 'styled-components'

export const LogoContainer = styled.div`
  width: 100%;
  position: absolute;
  display: flex;
  justify-content: ${props => props.position === 'right' ? 'flex-end' : 'flex-start'};
  align-items: flex-end;
  bottom: 1px;
  z-index: 1;
`

export const Logo = styled.a.attrs(props => ({
  title: props.title
}))`
  width: 15px;
  height: 15px;
  padding: 2px;
  cursor: pointer;
`

export const LogoText = styled.p`
  font-size: 11px;
  font-family: Roboto,Arial,sans-serif;
  padding-left: 2px;
  margin: 0px;
`

export const StyledMap = styled.div`
  height: 100%;
  width: 100%;
  position: ${p => p.fullScreen ? 'fixed' : 'relative'};
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  ${p => p.style};
`
