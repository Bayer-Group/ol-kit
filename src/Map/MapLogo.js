import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { LogoContainer, Logo, LogoText } from './styled'
import OL_KIT_MARK from 'images/ol_kit_mark.svg'
import OL_KIT_MARK_BLACK from 'images/ol_kit_mark_black.svg'

export default function MapLogo (props) {
  const { logoPosition, translations } = props
  const [isHovered, setHovered] = useState(false)

  return (
    logoPosition === 'none'
      ? null
      : (
        <LogoContainer position={logoPosition}>
          <LogoText>{translations['_ol_kit.MapLogo.title']}</LogoText>
          <Logo
            onMouseEnter={() => setHovered(true)}
            onMouseLeave={() => setHovered(false)}
            title={translations['_ol_kit.MapLogo.hover']}
            href='https://ol-kit.com/'
            target='_blank'>
            {isHovered
              ? <OL_KIT_MARK />
              : <OL_KIT_MARK_BLACK />
            }
          </Logo>
        </LogoContainer>
      )
  )
}

MapLogo.defaultProps = {
  logoPosition: 'right'
}

MapLogo.propTypes = {
  logoPosition: PropTypes.string,
  translations: PropTypes.object
}
