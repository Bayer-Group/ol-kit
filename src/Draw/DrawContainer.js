import React from 'react'
import PropTypes from 'prop-types'
import nanoid from 'nanoid'
import olLayerVector from 'ol/layer/Vector'
import olSourceVector from 'ol/source/Vector'
import { createBox } from 'ol/interaction/Draw'
import CircularProgress from '@material-ui/core/CircularProgress'
import Draw from './Draw'
import { Measure } from 'Measure'
import { SnapPreference, CoordinateLabelPreference } from 'Preferences'
import { connectToContext } from 'Provider'
import { styleMeasure } from './utils'
import { Container, ProgressWrapper } from './styled'

class MockPreferences {
  constructor () {
    this.state = {}
  }

  get (key) {
    return this.state[key]
  }

  async put (key, val) {
    this.state = { ...this.state, [key]: val }

    return val
  }
}

/**
 * A prebuilt Draw Tools component
 * @component
 * @category Draw
 * @since 0.18.0
 */
class DrawContainer extends React.Component {
  constructor () {
    super()

    this.state = {}

    this.onDrawStart = this.onDrawStart.bind(this)
    this.onDrawCancel = this.onDrawCancel.bind(this)
    this.onDrawEnd = this.onDrawEnd.bind(this)
    this.renderMeasure = this.renderMeasure.bind(this)
    this.handleToggle = this.handleToggle.bind(this)
  }

  safeGetPreference = (key) => this.props.preferences?.payload?.get?.(key)

  getUoms = () => {
    const uom = this.safeGetPreference('_UOM') || 'imperial'
    const distanceUOM = this.safeGetPreference('_DISTANCE_LABEL_UOM')
    const areaUOM = this.safeGetPreference('_AREA_LABEL_UOM')

    return {
      uomType: uom,
      distanceUOM: distanceUOM || (uom === 'imperial' ? 'feet' : 'meters'),
      areaUOM: areaUOM || (uom === 'imperial' ? 'acres' : 'hectares')
    }
  }

  updateMeasureStyle = (newValues = this.getUoms()) => {
    const { map } = this.props
    const { feature } = this.state
    const { distanceUOM, areaUOM } = newValues
    const isLegacyMeasure = feature?.get('_vmf_type') === '_vmf_measurement'
    const needsAreaLabels = feature?.get('_ol_kit_area_labels')
    const needsDistanceLabels = feature?.get('_ol_kit_distance_labels')
    const isMeasure = isLegacyMeasure || needsAreaLabels || needsDistanceLabels

    this.setState(newValues)

    if (isMeasure) {
      const styleFunc = styleMeasure.bind(this,
        map,
        feature,
        map.getView().getResolution(),
        { distanceUOM, areaUOM, map })

      feature.setStyle(styleFunc)
    }
  }

  getLayer () {
    const { feature } = this.state
    const title = feature.get('_vmf_type') === '_vmf_measurement' ? 'Measurements Layer' : 'Annotations Layer'
    const layers = this.props.map.getLayers().getArray()
    const exists = layers.find(l => l.get('_vmf_title') === title)

    if (exists) {
      return exists
    } else {
      const layer = new olLayerVector({
        className: `_ol_kit_${title}`,
        _vmf_id: nanoid(),
        _vmf_title: title,
        title: 'Annotations',
        source: new olSourceVector()
      })

      this.props.map.addLayer(layer)

      return layer
    }
  }

  onDrawEnd (feature) {
    console.debug('%cDrawEnd', 'color: cyan; font-style: italic;', feature) // eslint-disable-line

    const layer = this.getLayer()

    layer.getSource().addFeature(feature)
    feature.set('_ol_kit_parent', layer)
    feature.set('_ol_kit_annotation', true)
    feature.set('_vmf_id', nanoid())

    this.setState({ feature: null })
  }

  onDrawStart (feature, { target }) {
    const { map } = this.props
    const { distanceUOM, areaUOM } = this.getUoms()
    const pointLabels = this.safeGetPreference('_POINT_LABELS_ENABLED')
    const distanceLabelsEnabled = this.safeGetPreference('_DISTANCE_LABEL_ENABLED')
    const areaLabelsEnabled = this.safeGetPreference('_AREA_LABEL_ENABLED')
    const opts = { distanceUOM, areaUOM, map }
    const isBoxDraw = target.geometryFunction_?.toString() === createBox().toString()
    const drawMode = isBoxDraw ? 'Box' : target.mode_
    const isFreehand = target.freehand_

    if (distanceLabelsEnabled) feature.set('_ol_kit_distance_labels', true)
    if (areaLabelsEnabled) feature.set('_ol_kit_area_labels', true)
    if (pointLabels && !isFreehand) feature.set('_ol_kit_coordinate_labels', true)
    if (drawMode === 'Circle') {
      feature.set('_ol_kit_draw_mode', 'circle')

      if (pointLabels) {
        feature.set('_ol_kit_needs_centroid_label', true)
        feature.set('_ol_kit_coordinate_labels', false) // To prevent all needVertexLabels for running in styles.js
      }
    }
    const styleFunc = distanceLabelsEnabled || areaLabelsEnabled || pointLabels
      ? styleMeasure.bind(this, map, feature, map.getView().getResolution(), opts)
      : undefined

    feature.setStyle(styleFunc)

    this.setState({ feature, drawMode })
    console.debug('%cDrawStart', 'color: magenta; font-style: italic;') // eslint-disable-line
  }

  onInteractionAdded (interaction) {
    console.debug('%cInteractionAdded', 'color: yellow; font-style: italic;', interaction) // eslint-disable-line
  }

  onDrawCancel (interaction) {
    console.debug('%cDrawCancel', 'color: white; background-color: red; border-radius: 4px; font-style: italic;', interaction) // eslint-disable-line
    this.setState({ feature: null })
  }

  handleUomChange = () => {
    this.updateMeasureStyle()
  }

  selectedFeature = (feature) => {
    this.setState({ feature })
    this.props.selectedFeature(feature)
  }

  handleToggle = () => {
    this.forceUpdate()
  }

  renderPreferences = () => {
    const { translations, preferences } = this.props
    const { status, payload } = preferences

    switch (status) {
      case 'loading':
        return (
          <ProgressWrapper key={nanoid()}>
            <CircularProgress color={'primary'} />
          </ProgressWrapper>
        )
      case 'success':
        return (
          <React.Fragment key={'Snap'}>
            <SnapPreference
              translations={translations}
              preferences={payload}
              onChange={this.handleToggle}
              compact={false} />
          </React.Fragment>
        )
      default:
        return null
    }
  }

  renderMeasure = () => {
    const { uom, translations, showCoordinateLabels, preferences, showMeasurements } = this.props

    if (!showMeasurements) return null
    const { feature, geometryType, drawMode } = this.state
    const { status, payload } = preferences

    switch (status) {
      case 'loading':
        return (
          <ProgressWrapper key={nanoid()}>
            <CircularProgress color={'primary'} />
          </ProgressWrapper>
        )
      case 'success':
        return (
          <React.Fragment key={'Measure'}>
            {
              showCoordinateLabels
                ? <CoordinateLabelPreference
                  compact={true}
                  translations={translations}
                  preferences={payload}
                  onChange={this.handleToggle} />
                : null
            }
            <Measure
              {...this.props}
              feature={feature}
              uom={uom || payload?.get?.('_UOM')}
              onUomChange={this.handleUomChange}
              geometryType={geometryType}
              preferences={payload}
              drawMode={drawMode} />
          </React.Fragment>
        )
      default:
        return null
    }
  }

  render () {
    const { preferences, children, style } = this.props
    const drawChildren = children || [
      this.renderMeasure(),
      <Draw
        {...this.props}
        key={'Draw'}
        preferences={preferences.payload}
        onDrawFinish={this.onDrawEnd}
        onDrawBegin={this.onDrawStart}
        onInteractionAdded={this.onInteractionAdded}
        onDrawCancel={this.onDrawCancel}
        selectInteraction={this.props.selectInteraction} />,
      this.renderPreferences()
    ]

    return (
      <Container style={style}>
        {drawChildren}
      </Container>
    )
  }
}

DrawContainer.propTypes = {
  /** openlayers map */
  map: PropTypes.object.isRequired,
  /** reference to openlayers select interaction which can optionally be managed by IA */
  selectInteraction: PropTypes.object,
  /** an object with `get` and `put` methods to handle measure state */
  preferences: PropTypes.object.isRequired,
  /** translations object */
  translations: PropTypes.object,
  /** boolean for enabling snap interaction */
  snap: PropTypes.bool,
  /** openlayers snap options */
  snapOpts: PropTypes.object,
  /** boolean enabling the measurement component */
  showMeasurements: PropTypes.bool,
  /** boolean enabling the coordinate labels component */
  showCoordinateLabels: PropTypes.bool,
  /** velocity preference of either imperial or metric */
  uom: PropTypes.string,
  /** openlayers draw interaction constructor props */
  drawOpts: PropTypes.object,
  /** callback that returns the selected openlayers feature from the map */
  selectedFeature: PropTypes.func,
  /** pass custom style object to DrawContainer */
  style: PropTypes.object,
  /** pass child comps to opt out of the default controls */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
}

DrawContainer.defaultProps = {
  drawOpts: {},
  snap: true,
  showMeasurements: true,
  showCoordinateLabels: true,
  preferences: { status: 'success', payload: new MockPreferences() },
  selectedFeature: () => {}
}

export default connectToContext(DrawContainer)
