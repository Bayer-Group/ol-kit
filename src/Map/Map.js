import React from 'react'
import PropTypes from 'prop-types'
import nanoid from 'nanoid'
import { createMap } from './utils'

// context should only be created when <Map> is mounted (see constructor), otherwise it's null so child comps don't use context
export let MapContext = null

/**
 * A Reactified ol.Map wrapper component
 * @component
 * @category Map
 * @since 1.0.0
 */
class Map extends React.Component {
  constructor (props) {
    super(props)

    // this is used to create a unique identifier for the map div
    this.target = `_olKit_map_${nanoid(6)}`

    // create a map context
    MapContext = React.createContext()
  }

  componentDidMount () {
    const { map, onMapInit } = this.props

    if (!map) {
      // if no map was passed, create the map
      this.map = createMap({ target: this.target })
      // callback that returns a reference to the created map
      onMapInit(this.map)
      // force update to get this.map into the context
      this.forceUpdate()
    } else {
      this.map = map
    }
  }

  getContextValue = () => {
    return {
      map: this.map
    }
  }

  render () {
    return (
      <React.Fragment>
        <div id={this.target}></div>
        <MapContext.Provider value={this.getContextValue()}>
          {this.props.children}
        </MapContext.Provider>
      </React.Fragment>
    )
  }
}

Map.defaultProps = {
  map: null,
  onMapInit: () => {}
}

Map.propTypes = {
  /** all children components will automatically be passed a reference to the map object via the `map` prop */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  /** optionally pass a custom map */
  map: PropTypes.object,
  /** returns an initialized map object if nothing was passed to the `map` prop */
  onMapInit: PropTypes.func.isRequired
}

export default Map
