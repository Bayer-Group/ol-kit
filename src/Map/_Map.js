import React from 'react'
import Map from './Map' // this should be the only relative ./Map import in ol-kit
import { Provider } from 'Provider'

/* this becomes a ternary once <Provider> multi-map is available
   !ProviderContext
      ? (
        <Provider>
          <Map {...props} />
        </Provider>
      )
      : <Map {...props} />
*/
function _Map (props) {
  return (
    <Provider>
      <Map {...props} />
    </Provider>
  )
}

export default _Map
