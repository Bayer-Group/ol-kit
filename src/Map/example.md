# Map
The `<Map>` component will either create an [OpenLayers map](https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html) for you or take a prebuilt map as a prop. A reference to this map will be passed down via props to ol-kit child components of `<Map>` via [`connectToContext`](../connectToContext/example.md).

## Drop-in example:
This one-liner will create a map + div with the id of `_ol_kit_map_{uuid}` & render the map into that div.
```javascript static
import React from 'react'
import { Map } from '@bayer/ol-kit'

const App = () => {
  return (
    <Map />
  )
}

export default App
```


## Callback example:
`onMapInit` is not required but it will be called on [`componentDidMount`](https://reactjs.org/docs/react-component.html#componentdidmount) and return a reference to the newly created OpenLayers map (in case you want to manipulate it).
```javascript static
import React from 'react'
import { Map } from '@bayer/ol-kit'

import VectorLayer from 'ol/layer/Vector'
import VectorSource from 'ol/source/Vector'

class App extends React.Component {
  onMapInit = map => {
    const data = new VectorLayer({
      source: new VectorSource({
        features: [/** Get some data and have fun with it */]
      })
    })

    // add the data to the map
    map.addLayer(data)
  }

  render () {
    return (
      <Map onMapInit={this.onMapInit} />
    )
  }
}

export default App
```

## Prop example:
Make sure the passed map is created before `<Map>` is mounted; if the `map` prop is falsey on mount, `<Map>` will create it's own map instead (and return it via `onMapInit`).
```javascript static
import React from 'react'
import { Map, Popup, Controls } from '@bayer/ol-kit'

import Map from 'ol/Map'
import View from 'ol/View'
import TileLayer from 'ol/layer/Tile'
import OSM from 'ol/source/OSM'

class App extends React.Component {
  constructor () {
    super()

    // create new openlayers map
    const map = new Map({
      view: new View({
        center: [0, 0],
        zoom: 5
      }),
      layers: [
        new TileLayer({
          source: new OSM()
        })
      ],
      target: 'map'
    })

    this.state = {
      map
    }
  }

  render () {
    const { map } = this.state

    // create a div for the map to render into with an id matching the target
    return (
      <React.Fragment>
        <div id='map'></div>
        <Map map={map} />
      </React.Fragment>
    )
  }
}

export default App
```
