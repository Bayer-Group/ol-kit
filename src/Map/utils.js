import Map from 'ol/Map'
import View from 'ol/View'
import TileLayer from 'ol/layer/Tile'
import OSM from 'ol/source/OSM'

/**
 * Create an openlayers map
 * @function
 * @category Map
 * @since 1.0.0
 * @param {Object} [opts] - Object of optional params
 * @param {String} [opts.target] - htm id tag that map will into which the map will render
 * @returns {ol.Map} A newly constructed [ol.Map]{@link https://openlayers.org/en/v4.6.5/apidoc/ol.Map.html}
 */
export function createMap ({ target }) {
  // create a new map instance
  const map = new Map({
    view: new View({
      center: [0, 0],
      zoom: 5
    }),
    layers: [
      new TileLayer({
        source: new OSM()
      })
    ],
    target
  })

  return map
}


/**
 * An HOC designed to automatically pass down an ol.Map from the top-level Map component
 * @function
 * @category Map
 * @since 1.0.0
 * @param {Component} component - A React component you want wrapped
 * @returns {Component} A wrapped React component which will automatically be passed 
 */
export function MapConsumer (Component) {
  return props => {
    return (
      !MapContext
        ? <Component {...props} />
        : (
          <MapContext.Consumer>
            {({ map }) => <Component map={map} {...props} />}
          </MapContext.Consumer>
        )
    )
  }
}