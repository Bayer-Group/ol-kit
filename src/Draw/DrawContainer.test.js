import React from 'react'
import { render, waitFor } from '@testing-library/react'
import { prettyDOM } from '@testing-library/dom'
import { Map } from 'Map'
import DrawContainer from './DrawContainer'

describe('<DrawContainer />', () => {
  it('should render a basic prebuilt DrawContainer component', async () => {
    const { container } = render(<Map><DrawContainer /></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(prettyDOM(container)).toMatchSnapshot()
  })
  it.skip('should render a single child', async () => {
    const child = <div id='draw child comp'>draw child comp</div>
    const { container } = render(<Map><DrawContainer>{child}</DrawContainer></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(prettyDOM(container)).toMatchSnapshot()
  })
  it.skip('should render an array of children', async () => {
    const child1 = <div key={1} id='draw child comp 1'>child comp 1</div>
    const child2 = <div key={2} id='draw child comp 2'>child comp 2</div>
    const child3 = <div key={3} id='draw child comp 3'>child comp 3</div>
    const { container } = render(<Map><DrawContainer>{[child1, child2, child3]}</DrawContainer></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(prettyDOM(container)).toMatchSnapshot()
  })
})
