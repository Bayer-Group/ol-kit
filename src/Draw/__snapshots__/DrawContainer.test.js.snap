// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawContainer /> should render a basic prebuilt DrawContainer component 1`] = `
"[36m<div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-hKFyIo gQSimQ\\"[39m
    [33mid[39m=[32m\\"_ol_kit_map_test_id\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-bdnylx kHErEJ\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"sc-dlnjPT kmmhpS\\"[39m
      [36m>[39m
        [0mPowered by ol-kit[0m
      [36m</p>[39m
      [36m<a[39m
        [33mclass[39m=[32m\\"sc-gtssRu fBppYk\\"[39m
        [33mhref[39m=[32m\\"https://ol-kit.com/\\"[39m
        [33mtarget[39m=[32m\\"_blank\\"[39m
        [33mtitle[39m=[32m\\"Powered by ol-kit v1.4.1\\"[39m
      [36m>[39m
        [36m<svg[39m
          [33mviewBox[39m=[32m\\"0 0 204.76 236.44\\"[39m
          [33mxmlns[39m=[32m\\"http://www.w3.org/2000/svg\\"[39m
        [36m>[39m
          [36m<defs>[39m
            [36m<style>[39m
              [0m.cls-1[0m
              [0m{[0m
              [0mfill:#231f20[0m
              [0m}[0m
            [36m</style>[39m
          [36m</defs>[39m
          [36m<g[39m
            [33mid[39m=[32m\\"Layer_2\\"[39m
          [36m>[39m
            [36m<path[39m
              [33mclass[39m=[32m\\"cls-1\\"[39m
              [33md[39m=[32m\\"M102.38 0L0 59.11v118.22l102.38 59.11 102.38-59.11V59.11zM106 15.64l87 50.26v102L80.49 104.09zM18 62.27l74.66-43.1-33.95 118L18 62.27zM43.8 189l-32-18.51V75.37l42.35 77.86zm10.49 6.06l7.82-27.17 25.12 46.19zm50.38 26.44l-37.93-69.67 10.4-36.13 106.4 60.3z\\"[39m
              [33mid[39m=[32m\\"logo_examples\\"[39m
            [36m/>[39m
          [36m</g>[39m
        [36m</svg>[39m
      [36m</a>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"ol-viewport\\"[39m
      [33mstyle[39m=[32m\\"position: relative; overflow: hidden; width: 100%; height: 100%;\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-unselectable ol-layers\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; width: 100%; height: 100%; z-index: 0; display: none;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer-stopevent\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-eCApGN kpkDGM\\"[39m
  [36m>[39m
    [36m<div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-iCoHVE ghVwUy\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"sc-fujyUd fSNpKS\\"[39m
          [33mid[39m=[32m\\"_POINT_LABELS_ENABLED\\"[39m
        [36m>[39m
          [36m<label[39m
            [33mclass[39m=[32m\\"sc-pNWxx kryAYs\\"[39m
            [33mfor[39m=[32m\\"_POINT_LABELS_ENABLED\\"[39m
          [36m>[39m
            [0mCoordinate Labels[0m
          [36m</label>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"MuiSwitch-root\\"[39m
          [36m>[39m
            [36m<span[39m
              [33maria-disabled[39m=[32m\\"false\\"[39m
              [33mclass[39m=[32m\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-16 MuiSwitch-switchBase WithStyles(ForwardRef(Switch))-switchBase-1 MuiSwitch-colorPrimary\\"[39m
            [36m>[39m
              [36m<span[39m
                [33mclass[39m=[32m\\"MuiIconButton-label\\"[39m
              [36m>[39m
                [36m<input[39m
                  [33mclass[39m=[32m\\"PrivateSwitchBase-input-19 MuiSwitch-input\\"[39m
                  [33mtype[39m=[32m\\"checkbox\\"[39m
                  [33mvalue[39m=[32m\\"false\\"[39m
                [36m/>[39m
                [36m<span[39m
                  [33mclass[39m=[32m\\"MuiSwitch-thumb\\"[39m
                [36m/>[39m
              [36m</span>[39m
              [36m<span[39m
                [33mclass[39m=[32m\\"MuiTouchRipple-root\\"[39m
              [36m/>[39m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"MuiSwitch-track WithStyles(ForwardRef(Switch))-track-3\\"[39m
            [36m/>[39m
          [36m</span>[39m
        [36m</div>[39m
      [36m</div>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"MuiContainer-root WithStyles(ForwardRef(Container))-root-32 MuiContainer-maxWidthLg\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"MuiContainer-root WithStyles(ForwardRef(Container))-root-41 MuiContainer-maxWidthLg\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mstyle[39m=[32m\\"width: 100%;\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"sc-iCoHVE ghVwUy\\"[39m
          [36m>[39m
            [36m<div[39m
              [33mclass[39m=[32m\\"sc-fujyUd fSNpKS\\"[39m
              [33mid[39m=[32m\\"_DISTANCE_LABEL_ENABLED\\"[39m
            [36m>[39m
              [36m<span[39m
                [33mclass[39m=[32m\\"MuiSwitch-root\\"[39m
              [36m>[39m
                [36m<span[39m
                  [33maria-disabled[39m=[32m\\"false\\"[39m
                  [33mclass[39m=[32m\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-16 MuiSwitch-switchBase WithStyles(ForwardRef(Switch))-switchBase-1 MuiSwitch-colorPrimary\\"[39m
                [36m>[39m
                  [36m<span[39m
                    [33mclass[39m=[32m\\"MuiIconButton-label\\"[39m
                  [36m>[39m
                    [36m<input[39m
                      [33mclass[39m=[32m\\"PrivateSwitchBase-input-19 MuiSwitch-input\\"[39m
                      [33mdata-testid[39m=[32m\\"geokit-MeasureLabelPreference-uomToggle\\"[39m
                      [33mtype[39m=[32m\\"checkbox\\"[39m
                      [33mvalue[39m=[32m\\"false\\"[39m
                    [36m/>[39m
                    [36m<span[39m
                      [33mclass[39m=[32m\\"MuiSwitch-thumb\\"[39m
                    [36m/>[39m
                  [36m</span>[39m
                  [36m<span[39m
                    [33mclass[39m=[32m\\"MuiTouchRipple-root\\"[39m
                  [36m/>[39m
                [36m</span>[39m
                [36m<span[39m
                  [33mclass[39m=[32m\\"MuiSwitch-track WithStyles(ForwardRef(Switch))-track-3\\"[39m
                [36m/>[39m
              [36m</span>[39m
            [36m</div>[39m
            [36m<div[39m
              [33mclass[39m=[32m\\"sc-jrsJCI cIQApe\\"[39m
            [36m>[39m
              [36m<span>[39m
                [0mDistance: [0m
                [0m [0m
                [0m[0m
              [36m</span>[39m
            [36m</div>[39m
            [36m<div[39m
              [33mclass[39m=[32m\\"MuiInputBase-root MuiInput-root Mui..."
`;
